core/src/main/java/tc/oc/pgm/flag/LegacyFlagBeamMatchModule.java

```diff
@@ -0,0 +1,263 @@
+  @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
+  public void onMatchStart(MatchStartEvent event) {
+    flags().filter(flag -> flag.getState() instanceof Spawned).forEach(flag -> retrackFlag(flag));
+  }
+
+  @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
+  public void onPlayerJoinMatch(PlayerJoinMatchEvent event) {
+    flags()
+        .filter(flag -> flag.getState() instanceof Spawned)
+        .forEach(flag -> trackFlag(flag, event.getPlayer()));
+  }
+
+  @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
+  public void onPlayerLeaveMatch(PlayerLeaveMatchEvent event) {
+    flags()
+        .filter(flag -> flag.getState() instanceof Spawned)
+        .forEach(flag -> untrackFlag(flag, event.getPlayer()));
+  }
+
+  @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
+  public void onFlagStateChange(FlagStateChangeEvent event) {
+    Flag flag = event.getFlag();
+    untrackFlag(flag);
+    if (event.getNewState() instanceof Spawned) {
+      trackFlag(flag);
+    }
+  }
+
+  private class UpdateTask implements Runnable {
+
+    private final Future<?> task;
+
+    private UpdateTask() {
+      this.task =
+          match
+              .getExecutor(MatchScope.RUNNING)
+              .scheduleAtFixedRate(this, UPDATE_DELAY, UPDATE_FREQUENCY, TimeUnit.MILLISECONDS);
+    }
+
+    public void stop() {
+      this.task.cancel(true);
+    }
+
+    @Override
+    public void run() {
+      ImmutableList.copyOf(beams.values()).forEach(flags -> flags.forEach((f, b) -> b.update()));
+    }
+  }
+
+  class Beam {
+
+    final Flag flag;
+    final Player bukkit;
+    final List<NMSHacks.FakeZombie> segments;
+
+    Beam(Flag flag, Player bukkit) {
+      this.flag = flag;
+      this.bukkit = bukkit;
+      this.segments =
+          range(0, 100) // ~100 blocks is the height which the particles appear to be reasonably
+              // visible (similar amount to amount closest to the flag)
+              .mapToObj(i -> new NMSHacks.FakeZombie(match.getWorld(), true, false))
+              .collect(Collectors.toList());
+      show();
+    }
+
+    Optional<Player> carrier() {
+      return Optional.ofNullable(
+          flag.getState() instanceof Carried
+              ? ((Carried) flag.getState()).getCarrier().getBukkit()
+              : null);
+    }
+
+    Optional<Location> location() {
+      if (!flag.getLocation().isPresent()) {
+        return Optional.empty();
+      }
+
+      Location location = flag.getLocation().get().clone();
+      location.setPitch(0);
+      return Optional.of(location);
+    }
+
+    ItemStack wool() {
+      return new ItemBuilder()
+          .material(Material.WOOL)
+          .enchant(Enchantment.DURABILITY, 1)
+          .color(flag.getDyeColor())
+          .build();
+    }
+
+    void show() {
+      if (carrier().map(carrier -> carrier.equals(bukkit)).orElse(false)) return;
```

Comment:
If I am understanding this correctly the carrying player can not see their own beam?

Reply:
I believe if the flag can be seen by the carrier then there might be a zombie in them which can block the vision.